quantum/action_tapping.c

```diff
@@ -107,12 +111,29 @@ void action_tapping_process(keyrecord_t record) {
 /* return true when key event is processed or consumed. */
 bool process_tapping(keyrecord_t *keyp) {
     keyevent_t event = keyp->event;
+#    if (defined(AUTO_SHIFT_ENABLE) && defined(RETRO_SHIFT)) || defined(TAPPING_TERM_PER_KEY) || defined(PERMISSIVE_HOLD_PER_KEY) || defined(TAPPING_FORCE_HOLD_PER_KEY) || defined(HOLD_ON_OTHER_KEYPRESS_PER_KEY)
+    uint16_t tapping_keycode = get_record_keycode(&tapping_key, false);
+#    endif
 
     // if tapping
     if (IS_TAPPING_PRESSED()) {
-        if (WITHIN_TAPPING_TERM(event)) {
+        // clang-format off
+        if (WITHIN_TAPPING_TERM(event)
+#    if defined(AUTO_SHIFT_ENABLE) && defined(RETRO_SHIFT)
+            || (
+#        ifdef RETRO_TAPPING_PER_KEY
+                get_retro_tapping(tapping_keycode, keyp) &&
+#        endif
+                (RETRO_SHIFT + 0) != 0 && TIMER_DIFF_16(event.time, tapping_key.event.time) < (RETRO_SHIFT + 0)
```

Comment:
Is there a good reason why 0 is added to `RETRO_SHIFT` here and on line 363?

Reply:
One can `#define RETRO_SHIFT` without a value, which just makes holds without interrupting keypresses send the shifted key. Defining it to a value makes mod taps' hold actions trigger after that many ms, like the [default behavior](https://cdn.discordapp.com/attachments/451392568345755648/747388373781708930/tap-holds-qmk.pdf) with `TAPPING_TERM` to allow using mod taps with the mouse.

Adding zero makes the former compile. I suppose those conditionals could not have the `+0` and be in another `#if`, but there are quite enough preprocessor macros in this haha

Reply:
I assume you saw that when working on #15741 - I'll try to test that with Retro Shift (base Auto Shift won't have any issues) too soon, but if/when you have a go at it note that [you have to add mod/layer taps to Auto Shift.](https://github.com/qmk/qmk_firmware/issues/15458#issuecomment-1010448452)